source/extensions/filters/network/local_ratelimit/local_ratelimit.cc

```diff
@@ -10,26 +10,60 @@ namespace Extensions {
 namespace NetworkFilters {
 namespace LocalRateLimitFilter {
 
+SINGLETON_MANAGER_REGISTRATION(shared_local_ratelimit);
+
+LocalRateLimiterImplSharedPtr SharedRateLimitSingleton::get(
+    const envoy::extensions::filters::network::local_ratelimit::v3::LocalRateLimit& proto_config,
+    std::function<LocalRateLimiterImplSharedPtr()> create_fn) {
+  if (proto_config.share_key().empty()) {
+    // If there is no key, never share.
+    return create_fn();
+  }
+
+  const Key key{proto_config.share_key(), proto_config.token_bucket()};
+  auto it = limiters_.find(key);
+  if (it == limiters_.end()) {
+    ENVOY_LOG(debug, fmt::format("LocalRateLimit for share_key '{}' is creating a new token bucket "
+                                 "due to no match found",
+                                 proto_config.share_key()));
+    it = limiters_.emplace(key, create_fn()).first;
```

Comment:
I might be missing something here, but isn't this singleton shared across all worker threads?
If so, couldn't there be a race here where multiple emplace calls will be done with the same key?

Reply:
This code all runs at config-load time, which is always on the main thread. Adding `ASSERT_IS_MAIN_OR_TEST_THREAD` to make it clear and validate that it is true.